src/deserialize/parse.rs

```diff
@@ -0,0 +1,361 @@
+    let mut res: BTreeMap<String, Edn> = BTreeMap::new();
+    let mut key: Option<Edn> = None;
+    let mut val: Option<Edn> = None;
+    loop {
+        match chars.next() {
+            Some('}') => return Ok(Edn::Map(Map::new(res))),
+            Some(c) if !c.is_whitespace() && c != ',' => {
+                if key.is_some() {
+                    val = Some(parse(Some(c), chars)?);
+                } else {
+                    key = Some(parse(Some(c), chars)?);
+                }
+            }
+            Some(c) if c.is_whitespace() || c == ',' => (),
+            err => return Err(Error::ParseEdn(format!("{:?} could not be parsed", err))),
+        }
+
+        if key.is_some() && val.is_some() {
+            res.insert(key.unwrap().to_string(), val.unwrap());
+            key = None;
+            val = None;
+        }
+    }
+}
+
+use std::borrow::Cow;
+
+pub trait MaybeReplaceExt<'a> {
+    fn maybe_replace(self, find: &str, replacement: &str) -> Cow<'a, str>;
+}
+
+impl<'a> MaybeReplaceExt<'a> for &'a str {
+    fn maybe_replace(self, find: &str, replacement: &str) -> Cow<'a, str> {
+        if self.contains(find) {
+            self.replace(find, replacement).into()
+        } else {
+            self.into()
+        }
+    }
+}
+
+impl<'a> MaybeReplaceExt<'a> for Cow<'a, str> {
+    fn maybe_replace(self, find: &str, replacement: &str) -> Cow<'a, str> {
+        if self.contains(find) {
+            self.replace(find, replacement).into()
+        } else {
+            self
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::edn::{Double, Map, Set};
+    use crate::{map, set};
+
+    #[test]
+    fn parse_keyword() {
+        let mut key = ":keyword".chars();
+
+        assert_eq!(
+            parse_edn(key.next(), &mut key).unwrap(),
+            Edn::Key(":keyword".to_string())
+        )
+    }
+
+    #[test]
+    fn parse_str() {
+        let mut string = "\"hello world, from      RUST\"".chars();
+
+        assert_eq!(
+            parse_edn(string.next(), &mut string).unwrap(),
+            Edn::Str("hello world, from      RUST".to_string())
+        )
+    }
+
+    #[test]
+    fn parse_number() {
+        let mut uint = "143".chars();
+        let mut int = "-435143".chars();
+        let mut f = "-43.5143".chars();
+        let mut r = "43/5143".chars();
+        assert_eq!(parse_edn(uint.next(), &mut uint).unwrap(), Edn::UInt(143));
+        assert_eq!(parse_edn(int.next(), &mut int).unwrap(), Edn::Int(-435143));
+        assert_eq!(
+            parse_edn(f.next(), &mut f).unwrap(),
+            Edn::Double(Double::from(-43.5143))
+        );
+        assert_eq!(
+            parse_edn(r.next(), &mut r).unwrap(),
+            Edn::Rational("43/5143".to_string())
+        );
+    }
+
+    #[test]
+    fn parse_char() {
+        let mut c = "\\k".chars();
+
+        assert_eq!(parse_edn(c.next(), &mut c).unwrap(), Edn::Char('k'))
+    }
+
+    #[test]
+    fn parse_bool_or_nil() {
```

Comment:
I think we should test strings like `ture` and `flase` to check if it returns an error properly.